library/core/src/mem/valid_size.rs

```diff
@@ -0,0 +1,64 @@
+use crate::convert::TryFrom;
+use crate::{fmt, num};
+
+/// A type storing a possible object size (in bytes) in the rust abstract machine.
+///
+/// This can be thought of as a positive `isize`, or `usize` without the high bit
+/// set.  This is important because [`pointer::offset`] is UB for *byte* sizes
+/// too large for an `isize`, and there's a corresponding language limit on the
+/// size of any allocated object.
+///
+/// Note that particularly large sizes, while representable in this type, are
+/// likely not to be supported by actual allocators and machines.
+#[derive(Copy, Clone, Hash, Eq, PartialEq, Ord, PartialOrd)]
+#[repr(transparent)]
+#[cfg_attr(target_pointer_width = "16", rustc_layout_scalar_valid_range_end(0x7FFF))]
+#[cfg_attr(target_pointer_width = "32", rustc_layout_scalar_valid_range_end(0x7FFF_FFFF))]
+#[cfg_attr(target_pointer_width = "64", rustc_layout_scalar_valid_range_end(0x7FFF_FFFF_FFFF_FFFF))]
+pub(crate) struct ValidSize(usize);
```

Comment:
Would it not be simpler to just use `#[rustc_layout_scalar_valid_range_start(0)] struct ValidSize(isize)`?

Reply:
I remember someone (you?) mentioning that, and actually tried it first.  But [it doesn't work](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2bc2b8bd7983ca44779590da85944894) -- `isize`'s range is also `0 ..= 0xFF…FF`, as can be seen by the fact that [`NonZeroI32` only has `rustc_layout_scalar_valid_range_start(1)`](https://github.com/rust-lang/rust/blob/master/library/core/src/num/nonzero.rs#L40), without customizing the end.

Reply:
Ah right, the attributes are the byte range and don't care about if the target is signed. Which is also important for `NonZeroINN` which would otherwise have a middle niche rather than an end niche.